neldermead.go

```diff
@@ -0,0 +1,292 @@
+	reflectedValue float64    // Value at the last reflection point
+}
+
+func (n *NelderMead) Init(loc *Location, f *FunctionInfo, xNext []float64) (EvaluationType, IterationType, error) {
+	dim := len(loc.X)
+	if cap(n.vertices) < dim+1 {
+		n.vertices = make([][]float64, dim+1)
+	}
+	n.vertices = n.vertices[:dim+1]
+	for i := range n.vertices {
+		n.vertices[i] = resize(n.vertices[i], dim)
+	}
+	n.values = resize(n.values, dim+1)
+	n.centroid = resize(n.centroid, dim)
+	n.reflectedPoint = resize(n.reflectedPoint, dim)
+
+	// Default parameter choices are chosen in a dimension-dependent way
+	// from http://www.webpages.uidaho.edu/~fuchang/res/ANMS.pdf
+	if n.Reflection == 0 {
+		n.Reflection = 1
+	}
+	if n.Expansion == 0 {
+		n.Expansion = 1 + 2.0/float64(dim)
+	}
+	if n.Contraction == 0 {
+		n.Contraction = 0.75 - 1.0/(2.0*float64(dim))
+	}
+	if n.Shrink == 0 {
+		n.Shrink = 1 - 1.0/float64(dim)
+	}
+
+	if n.InitialVertices != nil {
+		// Initial simplex provided. Copy the locations and values, and sort them.
+		if len(n.InitialVertices) != dim+1 {
+			panic("neldermead: incorrect number of vertices in initial simplex")
+		}
+		if len(n.InitialValues) != dim+1 {
+			panic("neldermead: incorrect number of values in initial simplex")
+		}
+		for i := range n.InitialVertices {
+			if len(n.InitialVertices[i]) != dim {
+				panic("neldermead: vertex size mismatch")
+			}
+			copy(n.vertices[i], n.InitialVertices[i])
+		}
+		copy(n.values, n.InitialValues)
+		sort.Sort(nmVertexSorter{n.vertices, n.values})
+		computeCentroid(n.vertices, n.centroid)
+		return n.returnNext(nmReflected, xNext)
+	}
+
+	// No simplex provided. Begin initializing initial simplex. First simplex
+	// entry is the initial location, then step 1 in every direction.
+	copy(n.vertices[dim], loc.X)
+	n.values[dim] = loc.F
+	n.fillIdx = 0
+	copy(xNext, loc.X)
+	xNext[0] += 1
+	copy(n.vertices[0], xNext)
+	n.lastIter = nmInitialize
+	return FuncEvaluation, InitIteration, nil
+}
+
+// computeCentroid computes the centroid of all the simplex vertices except the
+// final one
+func computeCentroid(vertices [][]float64, centroid []float64) {
+	dim := len(centroid)
+	for i := range centroid {
+		centroid[i] = 0
+	}
+	for i := 0; i < dim; i++ {
+		vertex := vertices[i]
+		for j, v := range vertex {
+			centroid[j] += v
+		}
+	}
+	for i := range centroid {
+		centroid[i] /= float64(dim)
+	}
+}
+
+func (n *NelderMead) Iterate(loc *Location, xNext []float64) (EvaluationType, IterationType, error) {
+	dim := len(xNext)
+	switch n.lastIter {
+	case nmInitialize:
+		n.values[n.fillIdx] = loc.F
+		n.fillIdx++
+		if n.fillIdx == dim {
+			// Successfully finished building initial simplex.
+			sort.Sort(nmVertexSorter{n.vertices, n.values})
+			computeCentroid(n.vertices, n.centroid)
+			return n.returnNext(nmReflected, xNext)
+		}
+		copy(xNext, loc.X)
+		xNext[n.fillIdx-1] -= 1
+		xNext[n.fillIdx] += 1
+		copy(n.vertices[n.fillIdx], xNext)
```

Comment:
I'm moving it, but this code right now assumes specific values in xNext, which is bad, so I'm also fixing that.